### PR TITLE
feature/Alarm Notification

### PR DESCRIPTION
**Pull request by kiennguyensys**
### Summary:
- [Feature]
- Push Notification only works when it's raining outside
- Scheduled Time with AlarmManager

### Why need create this PR:
- Add feature

### Effect:
- None

### Results:
- 
Screenshot_1621591524
Screenshot_1621590907
-
Screenshot_1621563484
-
Screenshot_1621563503

### Link Issue:
https://jobchat.vn/job/-LPOJhtoknDaKAnQF91f/#73

### CODE REVIEW

**Comment by kiennguyensys** [comment]
@vuphu-jc
@chuongdevJC
@AnhJobChat
review e với mấy a. Em mời như này được chưa nhỉ

File: app/src/main/java/com/yoshitoke/weatheringwithyou/alarm/AlarmScheduler.kt [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+        // get the AlarmManager reference
+        val alarmMgr = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
+
+        // Schedule the alarms based on the days to administer the medicine
+        val days = context.resources.getStringArray(R.array.days)
+        if (alarmData.days != null) {
+            for (index in alarmData.days!!.indices) {
+
+                val day = alarmData.days!![index]
+                if (day != null) {
```
Comment (vuphu-jc) [inline app/src/main/java/com/yoshitoke/weatheringwithyou/alarm/AlarmScheduler.kt:37]: Don't force wrap !! in any cases.

File: app/src/main/java/com/yoshitoke/weatheringwithyou/alarm/AlarmScheduler.kt [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    fun removeAlarmsForReminder(context: Context, alarmData: AlarmData) {
+        val intent = Intent(context.applicationContext, AlarmReceiver::class.java)
+        intent.action = context.getString(R.string.action_notify_weather_alert)
+        intent.putExtra("id", alarmData.id)
+
+        // type must be unique so Intent.filterEquals passes the check to make distinct PendingIntents
+        // Schedule the alarms based on the days to administer the medicine
+        if (alarmData.days != null) {
+            for (i in alarmData.days!!.indices) {
+                val day = alarmData.days!![i]
```
Comment (vuphu-jc) [inline app/src/main/java/com/yoshitoke/weatheringwithyou/alarm/AlarmScheduler.kt:143]: Remove !!

File: app/src/main/java/com/yoshitoke/weatheringwithyou/mvp/view/AlarmResultFragment.kt [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        presenter.setResultFragmentView(this)
+        edit_btn.setOnClickListener{
+            presenter.onEditedPressed()
+        }
+    }
+
+    override fun onAlarmSet(alarmData: AlarmData) {
+        val hourText = "Hour: " + alarmData.hour.toString()
+        val minuteText = "Minute: " + alarmData.minute.toString()
+        val daysText = "Every: " + alarmData.days.toString()
```
Comment (vuphu-jc) [inline app/src/main/java/com/yoshitoke/weatheringwithyou/mvp/view/AlarmResultFragment.kt:42]: strings.xml

File: app/src/main/java/com/yoshitoke/weatheringwithyou/mvp/presenter/AlarmPresenter.kt [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    var mDays: List<String>? = null
+    var mHour: Int? = null
+    var mMinute: Int? = null
+    var mWeatherList: List<String>? = null
+    var alarmData: AlarmData? = null
+    var finalResultView: AlarmContract.View.FinalResultView? = null
+
+    override fun destroy() {
+        TODO("Not yet implemented")
+    }
```
Comment (vuphu-jc) [inline app/src/main/java/com/yoshitoke/weatheringwithyou/mvp/presenter/AlarmPresenter.kt:26]:
```suggestion
    override fun destroy() = Unit;
```

File: app/src/main/java/com/yoshitoke/weatheringwithyou/mvp/model/AlarmDatabaseHandler.kt [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        return ContentValues().apply {
+            put(ALARM_HOUR, alarmData.hour)
+            put(ALARM_MINUTE, alarmData.minute)
+            val stringBuilder: StringBuilder
+            if (alarmData.days != null) {
+                stringBuilder = StringBuilder()
+                for (i in alarmData.days!!.indices) {
+                    if (alarmData.days!![i] == null) {
+                        continue
+                    }
```
Comment (vuphu-jc) [inline app/src/main/java/com/yoshitoke/weatheringwithyou/mvp/model/AlarmDatabaseHandler.kt:49]: Remove !!

File: app/src/main/java/com/yoshitoke/weatheringwithyou/mvp/model/AlarmDatabaseHandler.kt [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    fun updateAlarm(alarmData: AlarmData): String {
+        val selectionArs = arrayOf(alarmData.id.toString())
+        val contentValues = createContentValues(alarmData)
+
+        val i = sqlObj.update(tableName, contentValues, "id=?", selectionArs)
+        if (i > 0) {
+            return "ok";
+        } else {
+            return "error";
+        }
```
Comment (vuphu-jc) [inline app/src/main/java/com/yoshitoke/weatheringwithyou/mvp/model/AlarmDatabaseHandler.kt:147]: return a boolean is better

File: app/src/main/java/com/yoshitoke/weatheringwithyou/mvp/model/AlarmDatabaseHandler.kt [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+        val selectionArs = arrayOf(id.toString())
+
+        val i = sqlObj.delete(tableName, "id=?", selectionArs)
+        if (i > 0) {
+            return "ok";
+        } else {
+
+            return "error";
+        }
```
Comment (vuphu-jc) [inline app/src/main/java/com/yoshitoke/weatheringwithyou/mvp/model/AlarmDatabaseHandler.kt:160]: return a boolean

File: app/src/main/java/com/yoshitoke/weatheringwithyou/mvp/model/AlarmDatabaseHandler.kt [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+                + ALARM_MINUTE + " INTEGER, "
+                + ALARM_DAYS + " TEXT NOT NULL, "
+                + ALARM_ADMINISTERED + " INTEGER DEFAULT 0);")
+
+        sqlite?.execSQL(CREATE_TABLE);
+    }
+
+    override fun onUpgrade(sqlite: SQLiteDatabase?, p1: Int, p2: Int) {
+
+        sqlite?.execSQL("Drop table IF EXISTS " + tableName)
```
Comment (vuphu-jc) [inline app/src/main/java/com/yoshitoke/weatheringwithyou/mvp/model/AlarmDatabaseHandler.kt:106]: Uppercase SQL query
```suggestion
        sqlite?.execSQL("DROP TABLE IF EXISTS " + tableName)
```

File: app/src/main/java/com/yoshitoke/weatheringwithyou/mvp/model/AlarmDatabaseHandler.kt [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+            }
+            val typeStringBuilder: StringBuilder
+            if (alarmData.weatherTypes != null) {
+                typeStringBuilder = StringBuilder()
+                for (i in alarmData.weatherTypes!!.indices) {
+                    typeStringBuilder.append(alarmData.weatherTypes!![i])
+                    if (i < alarmData.weatherTypes!!.size - 1) {
+                        typeStringBuilder.append("-")
+                    }
+                }
```
Comment (vuphu-jc) [inline app/src/main/java/com/yoshitoke/weatheringwithyou/mvp/model/AlarmDatabaseHandler.kt:65]: .

File: app/src/main/java/com/yoshitoke/weatheringwithyou/mvp/AlarmContract.kt [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+        fun onNextButtonPressed()
+
+        fun checkPreviousSetting()
+
+        fun onEditedPressed()
+
+        fun setResultFragmentView(fragment: View.FinalResultView)
+
+    }
+}
```
Comment (vuphu-jc) [inline app/src/main/java/com/yoshitoke/weatheringwithyou/mvp/AlarmContract.kt:36]: Logic of Presenter should be separated from Android. Don't use any words related to Android such as Button, EditText, Fragment, Pressed, ....

File: app/src/main/java/com/yoshitoke/weatheringwithyou/alarm/AlarmReceiver.kt [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+import android.util.Log
+import com.google.gson.Gson
+import com.yoshitoke.weatheringwithyou.R
+
+class AlarmReceiver : BroadcastReceiver() {
+
+    private val TAG = AlarmReceiver::class.java.simpleName
+
+    override fun onReceive(context: Context?, intent: Intent?) {
+        Log.d(TAG, "onReceive() called with: context = [$context], intent = [$intent]")
```
Comment (chuongdevJC) [inline app/src/main/java/com/yoshitoke/weatheringwithyou/alarm/AlarmReceiver.kt:15]: remove Log.d

File: app/src/main/java/com/yoshitoke/weatheringwithyou/alarm/BootReceiver.kt [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+import android.content.BroadcastReceiver
+import android.content.Context
+import android.content.Intent
+import com.yoshitoke.weatheringwithyou.mvp.model.AlarmDatabaseHandler
+import io.reactivex.android.schedulers.AndroidSchedulers
+
+class BootReceiver : BroadcastReceiver() {
+
+    override fun onReceive(context: Context?, intent: Intent?) {
+        if (context != null && intent?.action.equals("android.intent.action.BOOT_COMPLETED")) {
```
Comment (chuongdevJC) [inline app/src/main/java/com/yoshitoke/weatheringwithyou/alarm/BootReceiver.kt:12]: android.intent.action.BOOT_COMPLETED -> define constant

File: app/src/main/java/com/yoshitoke/weatheringwithyou/alarm/WeatherAlertHelper.kt [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+                        {result ->
+                            val currentWeather = result.current.weathers[0].main
+                            val list = alarmData.weatherTypes as List<String>
+                            Log.d("weatherAlert: ", list.toString())
+
+                            if (list.contains(currentWeather)) {
+                                NotificationHelper.createSampleDataNotification(context, "Alert", "fuck", "Current Weather Outside: $currentWeather", false)
+                            }
+                        },
+                        {error -> Log.d("api error", error.message.toString())}
```
Comment (chuongdevJC) [inline app/src/main/java/com/yoshitoke/weatheringwithyou/alarm/WeatherAlertHelper.kt:27]: Log.e instead

File: app/src/main/java/com/yoshitoke/weatheringwithyou/alarm/WeatherAlertHelper.kt [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+                .subscribeOn(Schedulers.io())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(
+                        {result ->
+                            val currentWeather = result.current.weathers[0].main
+                            val list = alarmData.weatherTypes as List<String>
+                            Log.d("weatherAlert: ", list.toString())
+
+                            if (list.contains(currentWeather)) {
+                                NotificationHelper.createSampleDataNotification(context, "Alert", "fuck", "Current Weather Outside: $currentWeather", false)
```
Comment (chuongdevJC) [inline app/src/main/java/com/yoshitoke/weatheringwithyou/alarm/WeatherAlertHelper.kt:24]: ?

File: app/src/main/java/com/yoshitoke/weatheringwithyou/alarm/WeatherAlertHelper.kt [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+    var bruneiTest: City = City("Brunei",  114.949008710682, 4.8977738654898)
+    fun checkCurrentWeatherForRegisteredAlarm(alarmData: AlarmData, context: Context) {
+        val disposable = RetrofitClient.weatherService.getWeatherInfo(bruneiTest.latitude.toString(), bruneiTest.longitude.toString())
+                .subscribeOn(Schedulers.io())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(
+                        {result ->
+                            val currentWeather = result.current.weathers[0].main
+                            val list = alarmData.weatherTypes as List<String>
+                            Log.d("weatherAlert: ", list.toString())
```
Comment (chuongdevJC) [inline app/src/main/java/com/yoshitoke/weatheringwithyou/alarm/WeatherAlertHelper.kt:21]: remove

File: app/src/main/java/com/yoshitoke/weatheringwithyou/alarm/AlarmScheduler.kt [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+     *
+     * @param context      current application context
+     * @param alarmData AlarmData to use for the alarm
+     */
+//    fun scheduleAlarmsForData(context: Context) {
+//        val alarmDataList = DataUtils.loadUnAdministeredReminders()
+//        for (alarmData in alarmDataList) {
+//            AlarmScheduler.scheduleAlarmsForReminder(context, alarmData)
+//        }
+//    }
```
Comment (AnhJobChat) [inline app/src/main/java/com/yoshitoke/weatheringwithyou/alarm/AlarmScheduler.kt:24]: remove

File: app/src/main/java/com/yoshitoke/weatheringwithyou/alarm/AlarmReceiver.kt [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+        if (context != null && intent != null && intent.action != null) {
+            val json = intent.getStringExtra("alarm")
```
Comment (AnhJobChat) [inline app/src/main/java/com/yoshitoke/weatheringwithyou/alarm/AlarmReceiver.kt:17]: define it into extra constant

File: app/src/main/java/com/yoshitoke/weatheringwithyou/alarm/AlarmScheduler.kt [hunk 18]
```diff
@@ -0,0 +1,177 @@
...
+    /**
+     * Removes the notification if it was previously scheduled.
+     *
+     * @param context      current application context
+     * @param alarmData AlarmData for the notification
+     */
+    fun removeAlarmsForReminder(context: Context, alarmData: AlarmData) {
+        val intent = Intent(context.applicationContext, AlarmReceiver::class.java)
+        intent.action = context.getString(R.string.action_notify_weather_alert)
+        intent.putExtra("id", alarmData.id)
```
Comment (AnhJobChat) [inline app/src/main/java/com/yoshitoke/weatheringwithyou/alarm/AlarmScheduler.kt:137]: .

File: app/src/main/java/com/yoshitoke/weatheringwithyou/alarm/AlarmScheduler.kt [hunk 19]
```diff
@@ -0,0 +1,177 @@
...
+     * @param context      current application context
+     * @param alarmData AlarmData for the notification
+     * @param day          String representation of the day
+     */
+    private fun createPendingIntent(context: Context, alarmData: AlarmData, day: String?): PendingIntent? {
+        // create the intent using a unique type
+        val intent = Intent(context.applicationContext, AlarmReceiver::class.java).apply {
+            action = context.getString(R.string.action_notify_weather_alert)
+            type = "$day-${alarmData.weatherTypes}"
+            putExtra("alarm", Gson().toJson(alarmData))
```
Comment (AnhJobChat) [inline app/src/main/java/com/yoshitoke/weatheringwithyou/alarm/AlarmScheduler.kt:92]: .

File: app/src/main/java/com/yoshitoke/weatheringwithyou/alarm/WeatherAlertHelper.kt (hunk 14 above)
Comment (AnhJobChat) [inline app/src/main/java/com/yoshitoke/weatheringwithyou/alarm/WeatherAlertHelper.kt:24]: strings

File: app/src/main/java/com/yoshitoke/weatheringwithyou/mvp/view/AlarmActivity.kt [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+        screenPager.currentItem = screenPager.currentItem + 1
+    }
+
+    override fun onSpecificViewTransition(position: Int) {
+        screenPager.currentItem = position
+    }
+//    override fun onBackPressed() {
+//        setResult(RESULT_CANCELED)
+//        super.onBackPressed()
+//    }
```
Comment (AnhJobChat) [inline app/src/main/java/com/yoshitoke/weatheringwithyou/mvp/view/AlarmActivity.kt:42]: remove

File: app/src/main/java/com/yoshitoke/weatheringwithyou/mvp/view/AlarmPagerAdapter.kt [hunk 21]
```diff
@@ -0,0 +1,21 @@
+package com.yoshitoke.weatheringwithyou.mvp.view
+
+import androidx.appcompat.app.AppCompatActivity
+import androidx.fragment.app.Fragment
+import androidx.viewpager2.adapter.FragmentStateAdapter
+import com.yoshitoke.weatheringwithyou.mvp.presenter.AlarmPresenter
+
+class AlarmPagerAdapter(activity: AppCompatActivity, private val itemsCount: Int, private val presenter: AlarmPresenter) : FragmentStateAdapter(activity) {
```
Comment (AnhJobChat) [inline app/src/main/java/com/yoshitoke/weatheringwithyou/mvp/view/AlarmPagerAdapter.kt:8]:
```suggestion
class AlarmPagerAdapter(activity: AppCompatActivity, private val itemCount: Int, private val presenter: AlarmPresenter) : FragmentStateAdapter(activity) {
```

File: app/src/main/res/layout/fragment_alarm_result.xml [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:paddingHorizontal="@dimen/dp_20"
+    android:layout_height="match_parent">
+    <ImageView
+        android:id="@+id/notificationIcon"
+        android:layout_width="@dimen/dp_100"
+        android:layout_height="@dimen/dp_100"
+        android:layout_marginStart="@dimen/dp_16"
+        android:layout_marginTop="16dp"
```
Comment (AnhJobChat) [inline app/src/main/res/layout/fragment_alarm_result.xml:14]: .

File: app/src/main/res/layout/fragment_alarm_result.xml [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toBottomOf="@+id/notificationIcon" />
+
+
+
+    <Button
+        android:id="@+id/edit_btn"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:text="Edit"
```
Comment (AnhJobChat) [inline app/src/main/res/layout/fragment_alarm_result.xml:36]: .

File: app/src/main/res/layout/fragment_alarm_result.xml [hunk 24]
```diff
@@ -0,0 +1,71 @@
...
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent" />
+
+    <TextView
+        android:id="@+id/tv_hours"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginStart="@dimen/dp_16"
+        android:layout_marginTop="@dimen/dp_40"
+        android:text="TextView"
```
Comment (AnhJobChat) [inline app/src/main/res/layout/fragment_alarm_result.xml:47]: .

File: app/src/main/res/layout/fragment_alarm_result.xml [hunk 25]
```diff
@@ -0,0 +1,71 @@
...
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toBottomOf="@+id/textView6" />
+
+    <TextView
+        android:id="@+id/tv_days"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginStart="@dimen/dp_16"
+        android:layout_marginTop="@dimen/dp_32"
+        android:text="TextView"
```
Comment (AnhJobChat) [inline app/src/main/res/layout/fragment_alarm_result.xml:57]: .

File: app/src/main/res/layout/fragment_alarm_result.xml [hunk 26]
```diff
@@ -0,0 +1,71 @@
...
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toBottomOf="@+id/tv_hours" />
+
+    <TextView
+        android:id="@+id/tv_minute"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginStart="@dimen/dp_28"
+        android:layout_marginTop="@dimen/dp_40"
+        android:text="TextView"
```
Comment (AnhJobChat) [inline app/src/main/res/layout/fragment_alarm_result.xml:67]: .

File: app/src/main/res/layout/fragment_alarm_select_weather.xml [hunk 27]
```diff
@@ -0,0 +1,84 @@
...
+        android:singleLine="true"
+        android:text="@string/swelter"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toBottomOf="@+id/textView2" />
+
+    <Button
+        android:id="@+id/save_btn"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:text="Next"
```
Comment (AnhJobChat) [inline app/src/main/res/layout/fragment_alarm_select_weather.xml:80]: .

File: app/src/main/res/layout/fragment_alarm_set_time.xml [hunk 28]
```diff
@@ -0,0 +1,81 @@
...
+        app:layout_constraintLeft_toLeftOf="parent"
+        app:layout_constraintRight_toRightOf="parent"
+        app:layout_constraintTop_toBottomOf="@+id/textViewDaysLabel" />
+
+    <Button
+        android:id="@+id/next_btn"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginEnd="@dimen/dp_4"
+        android:text="Next"
```
Comment (AnhJobChat) [inline app/src/main/res/layout/fragment_alarm_set_time.xml:70]: .

File: app/src/main/res/layout/fragment_alarm_set_time.xml [hunk 29]
```diff
@@ -0,0 +1,81 @@
...
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintEnd_toEndOf="parent" />
+
+    <Button
+        android:id="@+id/button4"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:text="Previous"
```
Comment (AnhJobChat) [inline app/src/main/res/layout/fragment_alarm_set_time.xml:78]: .

File: app/src/main/res/layout/fragment_alarm_welcome.xml [hunk 30]
```diff
@@ -0,0 +1,40 @@
...
+        android:layout_marginTop="@dimen/dp_24"
+        android:background="@android:drawable/ic_popup_reminder"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent" />
+
+    <Button
+        android:id="@+id/start_btn"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:text="Start"
```
Comment (AnhJobChat) [inline app/src/main/res/layout/fragment_alarm_welcome.xml:24]: .

**Pull request closed without merging**